crates/redstone/src/core/aggregator.rs

```diff
@@ -6,370 +6,482 @@ use crate::{
-                .iter()
-                .map(|v| v.to_u256())
-                .collect::<Vec<_>>()
-                .median()
-                .ok_or(Error::ArrayIsEmpty)?;
-
-            Ok(Value::from_u256(median))
-        })
-        .collect()
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct FeedValue {
+    pub feed: FeedId,
+    pub value: Value,
 }
 
-/// Makes the value signer matrix.
-/// This function may fail if DataPackage contains DataPoints with reoccurring FeedId
-/// or if FeedId has a wrong ASCII representation,
-/// or the recovered signer is not recognized.
-/// Check FeedId crate for more details.
-fn make_value_signer_matrix(
+pub fn aggregate_values(
     config: &Config,
     data_packages: Vec<DataPackage>,
-) -> Result<Matrix, Error> {
-    let mut matrix = vec![vec![None; config.signers().len()]; config.feed_ids().len()];
+) -> Result<Vec<FeedValue>, Error> {
+    let feed_count = config.feed_ids().len();
+    let signer_count = config.signers().len();
+    let size = feed_count * signer_count;
+
+    let mut values = vec![None; size];
+
+    for data_package in data_packages {
+        if data_package.signer_address.is_invalid() {
+            continue;
+        }
 
-    for data_package in data_packages.iter() {
-        let signer_index = config
-            .signer_index(&data_package.signer_address)
-            .ok_or(Error::SignerNotRecognized(data_package.signer_address))?;
+        let signer_idx = match config.signer_index(&data_package.signer_address) {
+            Some(idx) => idx,
+            _ => continue,
+        };
 
-        'data_points_iter: for data_point in data_package.data_points.iter() {
-            let Some(feed_index) = config.feed_index(data_point.feed_id) else {
-                continue 'data_points_iter;
+        for data_point in data_package.data_points {
+            let feed_idx = match config.feed_index(data_point.feed_id) {
+                Some(idx) => idx,
+                _ => continue,
             };
-            if matrix[feed_index][signer_index].is_some() {
+
+            if values[feed_idx * signer_count + signer_idx].is_some() {
                 return Err(Error::ReoccurringFeedId(data_point.feed_id));
             }
-            matrix[feed_index][signer_index] = data_point.value.into();
+
+            values[feed_idx * signer_count + signer_idx] = Some(data_point.value);
+        }
+    }
+
+    let mut result = vec![];
+
+    for feed in 0..feed_count {
+        let feed_values: Vec<_> = values
+            .iter()
+            .skip(feed * signer_count)
+            .take(signer_count)
+            .flatten()
+            .map(|v| v.to_u256())
+            .collect();
+
+        if feed_values.len() < config.signer_count_threshold() as usize {
+            continue;
         }
+
+        let median = match feed_values.median() {
+            Some(median) => median,
+            _ => continue,
+        };
+
+        if median == 0 {
```

Comment:
should we skip only 0-median feeds or all 0-value data points?
For example, [0, 1000] would give us 500.